### PR TITLE
Architecture: Add Parent Blueprint Substate Access + Other Refactors

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Add ability for nested blueprints to access it's parent blueprint's substates.

### System API updates
* Add `lock_parent_field` to `ActorApi`
* Return back field substates on `drop_object` in `ObjectApi`
* Add `globalize_with_address_and_child_object` in `ObjectApi` which allows one to create an object with a nested object without requiring additional method calls
* Allow nested objects to create other nested objects if it's the same parent blueprint instance

### Blueprint updates
* Separate Fungible and NonFungible Proof implementations
* Separate out ResourceManager substates into more granular substates
* Separate out EpochManager substates into more granular substates
* Update Validator blueprint to use `lock_parent_field` api
* Update Bucket/Vault blueprint to use `lock_parent_field` api

### For Internal Integrators
* Separated out ResourceManager substates into more granular substates
* Separate dout EpochManager substates into more granular substates

### CODE REVIEW

File: radix-engine-interface/src/types/blueprint.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ use sbor::rust::string::ToString;
 pub struct ObjectInfo {
     pub blueprint: Blueprint,
     pub global: bool,
-    pub type_parent: Option<GlobalAddress>,
+    pub blueprint_parent: Option<GlobalAddress>,
```
Comment (iamyulong) [inline radix-engine-interface/src/types/blueprint.rs:11]: I'd suggest we use `inner`/`outer` (and reserve `child`/`parent` for ownership).

So rename to `outter_object`?

File: radix-engine-interface/src/api/object_api.rs [hunk 2]
```diff
@@ -55,6 +55,14 @@ pub trait ClientObjectApi<E> {
         address: GlobalAddress,
     ) -> Result<(), E>;
 
+    fn globalize_with_address_and_child_object(
```
Comment (iamyulong) [inline radix-engine-interface/src/api/object_api.rs:58]: I'd suggest we use `inner`/`outer` (and reserve `child`/`parent` for ownership).

So rename to `globalize_with_address_and_create_inner_object`?

File: radix-engine/src/blueprints/resource/package.rs [hunk 3]
```diff
@@ -978,71 +995,132 @@ impl ResourceManagerNativePackage {
...
-        let proof_schema = BlueprintSchema {
-            parent: None,
-            schema,
-            substates,
-            functions,
-            virtual_lazy_load_functions: btreemap!(),
-            event_schema: [].into(),
+            let schema = generate_full_schema(aggregator);
+            BlueprintSchema {
+                parent: Some(FUNGIBLE_RESOURCE_MANAGER_BLUEPRINT.to_string()),
```
Comment (iamyulong) [inline radix-engine/src/blueprints/resource/package.rs:1047]: Rename `parent` to `outer_blueprint`?

File: radix-engine/src/system/system_modules/auth/auth_module.rs [hunk 4]
```diff
@@ -293,7 +294,7 @@ impl AuthModule {
                     .child_blueprint_rules
                     .get(&blueprint_name)
                     .ok_or(RuntimeError::ModuleError(ModuleError::AuthError(
-                        AuthError::ChildBlueprintDoesNotExist,
+                        AuthError::ChildBlueprintDoesNotExist(blueprint_name),
```
Comment (iamyulong) [inline radix-engine/src/system/system_modules/auth/auth_module.rs:297]: `child_blueprint` => `inner_blueprint`?

**Review by iamyulong — COMMENTED** [review]
Looks good! All my comments are about naming. :)

File: radix-engine-interface/src/types/blueprint.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine-interface/src/types/blueprint.rs:11]: cool sounds good

File: radix-engine-interface/src/api/object_api.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine-interface/src/api/object_api.rs:58]: yup makes sense

**Review by iamyulong — APPROVED** [review]

**Pull request merged**